The IBM database fails with Error: Unprocessable Entity

Steps 1 :

1) Create a database using `members_XXXX_XXX` (the values here for all members) argumets
```
resource "ibm_database" "database_instance" {
  name              = "test"
  service           = "databases-for-postgresql"
  plan              = "standard"
  location          = "eu-de"   
  resource_group_id = "XXXXXXXXX"
  plan_validation   = false
  
  adminpassword                = var.admin_password
  members_cpu_allocation_count = 0
  members_disk_allocation_mb = 10240
  members_memory_allocation_mb = 2048
  
 timeouts {
    create = "1440m"
    update = "120m"
    delete = "60m"
  } 

  lifecycle {
    ignore_changes = [
      # Ignore changes to tags, e.g. because a management agent
      # updates these based on some ruleset managed elsewhere.
      adminpassword,
    ]
  } 

}

```
<img width="762" alt="Screenshot 2022-08-11 at 12 54 33 PM" src="https://user-images.githubusercontent.com/24930306/184083850-a8eb5987-f0b8-4ac3-a9dc-0de76b640747.png">

2) Run terraform plan/apply.  After successfull provison change the tempalte to use `groups`-argument
```
resource "ibm_database" "database_instance" {
  name              = "test"
  service           = "databases-for-postgresql"
  plan              = "standard"
  location          = "eu-de"   
  resource_group_id = "XXXXXXXXX"
  plan_validation   = false
  
  adminpassword                = var.admin_password
  /*members_cpu_allocation_count = 0
  members_disk_allocation_mb = 10240
  members_memory_allocation_mb = 2048*/
  
  group {
    group_id = "member"

    memory { 
      allocation_mb = "1024" // The value here should be half of the members_memory_allocation_mb since its per member"
    }

    disk { 
      allocation_mb = "5120" // // The value here should be half of the members_disk_allocation_mb since its per member"
    }

    cpu {
      allocation_count = "0"
    }
  }
  
 timeouts {
    create = "1440m"
    update = "120m"
    delete = "60m"
  } 

  lifecycle {
    ignore_changes = [
      # Ignore changes to tags, e.g. because a management agent
      # updates these based on some ruleset managed elsewhere.
      adminpassword,
    ]
  } 

}
```
4 Now run terraform plan/apply it fails with 
```
22/08/11 14:08:51 [DEBUG] PATCH https://api.eu-de.databases.cloud.ibm.com/v5/ibm/deployments/crn:v1:bluemix:public:databases-for-postgresql:eu-de:a%!F(MISSING)1f7277194bb748cdb1d35fd8fb85a7cb:4dfe7cbe-2699-4cea-935b-f6bbe1df7766::/groups/member: timestamp=2022-08-11T14:08:51.276+0530
2022-08-11T14:08:51.895+0530 [INFO]  provider.terraform-provider-ibm_v1.42.0: 2022/08/11 14:08:51 [Debug] Response:
HTTP/2.0 422 Unprocessable Entity
Cache-Control: no-cache
Cf-Cache-Status: DYNAMIC
Cf-Ray: 738fa84b3c6da92c-MAA
Content-Security-Policy: frame-ancestors *.bluemix.net *.compose.fastkit *.cloud.ibm.com cloud.ibm.com
Content-Type: application/json; charset=utf-8
Date: Thu, 11 Aug 2022 08:38:51 GMT
Expect-Ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Referrer-Policy: strict-origin-when-cross-origin
Server: cloudflare
Strict-Transport-Security: max-age=0; includeSubDomains
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Envoy-Upstream-Service-Time: 431
X-Permitted-Cross-Domain-Policies: none
X-Request-Id: e03e4d41-f319-4763-8272-f30c6d302216
X-Runtime: 0.426730
X-Xss-Protection: 1; mode=block

{"errors":{"group":["is already at that value"]}}: timestamp=2022-08-11T14:08:51.894+0530
2022-08-11T14:08:51.925+0530 [ERROR] vertex "ibm_database.database_instance" error: Unprocessable Entity
╷
│ Error: Unprocessable Entity
│ 
│   with ibm_database.database_instance,
│   on database.tf line 5, in resource "ibm_database" "database_instance":
│    5: resource "ibm_database" "database_instance" {
│ 
╵

```
It fails with this error because already members are having same values we just migrated to use groups. This scenario needs to be handle in Terraform not to throw error just update the state file with groups argument

--- a/.secrets.baseline
+++ b/.secrets.baseline
@@ -3,7 +3,7 @@
     "files": "go.sum|^.secrets.baseline$|metadata",
     "lines": null
   },
-  "generated_at": "2022-08-18T09:44:33Z",
+  "generated_at": "2022-08-24T16:50:18Z",
   "plugins_used": [
     {
       "name": "AWSKeyDetector"
@@ -1720,7 +1720,7 @@
         "hashed_secret": "10c28f9cf0668595d45c1090a7b4a2ae98edfa58",
         "is_secret": false,
         "is_verified": false,
-        "line_number": 729,
+        "line_number": 779,
         "type": "Secret Keyword",
         "verified_result": null
       }
@@ -1778,7 +1778,7 @@
         "hashed_secret": "10c28f9cf0668595d45c1090a7b4a2ae98edfa58",
         "is_secret": false,
         "is_verified": false,
-        "line_number": 825,
+        "line_number": 871,
         "type": "Secret Keyword",
         "verified_result": null
       }
@@ -1798,7 +1798,7 @@
         "hashed_secret": "10c28f9cf0668595d45c1090a7b4a2ae98edfa58",
         "is_secret": false,
         "is_verified": false,
-        "line_number": 212,
+        "line_number": 243,
         "type": "Secret Keyword",
         "verified_result": null
       }

--- a/ibm/service/database/resource_ibm_database.go
+++ b/ibm/service/database/resource_ibm_database.go
@@ -206,7 +206,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"node_count", "node_memory_allocation_mb", "node_disk_allocation_mb", "node_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"members_disk_allocation_mb": {
 				Description:   "Disk allocation required for cluster",
@@ -214,7 +214,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"node_count", "node_memory_allocation_mb", "node_disk_allocation_mb", "node_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"members_cpu_allocation_count": {
 				Description:   "CPU allocation required for cluster",
@@ -222,7 +222,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"node_count", "node_memory_allocation_mb", "node_disk_allocation_mb", "node_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"node_count": {
 				Description:   "Total number of nodes in the cluster",
@@ -230,7 +230,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"members_memory_allocation_mb", "members_disk_allocation_mb", "members_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"node_memory_allocation_mb": {
 				Description:   "Memory allocation per node",
@@ -238,7 +238,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"members_memory_allocation_mb", "members_disk_allocation_mb", "members_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"node_disk_allocation_mb": {
 				Description:   "Disk allocation per node",
@@ -246,7 +246,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"members_memory_allocation_mb", "members_disk_allocation_mb", "members_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"node_cpu_allocation_count": {
 				Description:   "CPU allocation per node",
@@ -254,7 +254,7 @@ func ResourceIBMDatabaseInstance() *schema.Resource {
 				Optional:      true,
 				Computed:      true,
 				ConflictsWith: []string{"members_memory_allocation_mb", "members_disk_allocation_mb", "members_cpu_allocation_count", "group"},
-				Deprecated:    "This field is deprecated please use groups",
+				Deprecated:    "use group instead",
 			},
 			"plan_validation": {
 				Description: "For elasticsearch and postgres perform database parameter validation during the plan phase. Otherwise, database parameter validation happens in apply phase.",
@@ -1363,14 +1363,14 @@ func resourceIBMDatabaseInstanceCreate(context context.Context, d *schema.Resour
 				continue
 			}
 
-			if g.Members != nil && g.Members.Allocation != currentGroup.Members.Allocation {
+			if g.Members != nil && g.Members.Allocation*nodeCount != currentGroup.Members.Allocation {
 				groupScaling.Members = &clouddatabasesv5.GroupScalingMembers{AllocationCount: core.Int64Ptr(int64(g.Members.Allocation))}
 				nodeCount = g.Members.Allocation
 			}
-			if g.Memory != nil && g.Memory.Allocation != currentGroup.Memory.Allocation {
+			if g.Memory != nil && g.Memory.Allocation*nodeCount != currentGroup.Memory.Allocation {
 				groupScaling.Memory = &clouddatabasesv5.GroupScalingMemory{AllocationMb: core.Int64Ptr(int64(g.Memory.Allocation * nodeCount))}
 			}
-			if g.Disk != nil && g.Disk.Allocation != currentGroup.Disk.Allocation {
+			if g.Disk != nil && g.Disk.Allocation*nodeCount != currentGroup.Disk.Allocation {
 				groupScaling.Disk = &clouddatabasesv5.GroupScalingDisk{AllocationMb: core.Int64Ptr(int64(g.Disk.Allocation * nodeCount))}
 			}
 			if g.CPU != nil && g.CPU.Allocation != currentGroup.CPU.Allocation {
@@ -1926,14 +1926,14 @@ func resourceIBMDatabaseInstanceUpdate(context context.Context, d *schema.Resour
 			}
 			nodeCount := currentGroup.Members.Allocation
 
-			if group.Members != nil && group.Members.Allocation != currentGroup.Members.Allocation {
+			if group.Members != nil && group.Members.Allocation*nodeCount != currentGroup.Members.Allocation {
 				groupScaling.Members = &clouddatabasesv5.GroupScalingMembers{AllocationCount: core.Int64Ptr(int64(group.Members.Allocation))}
 				nodeCount = group.Members.Allocation
 			}
-			if group.Memory != nil && group.Memory.Allocation != currentGroup.Memory.Allocation {
+			if group.Memory != nil && group.Memory.Allocation*nodeCount != currentGroup.Memory.Allocation {
 				groupScaling.Memory = &clouddatabasesv5.GroupScalingMemory{AllocationMb: core.Int64Ptr(int64(group.Memory.Allocation * nodeCount))}
 			}
-			if group.Disk != nil && group.Disk.Allocation != currentGroup.Disk.Allocation {
+			if group.Disk != nil && group.Disk.Allocation*nodeCount != currentGroup.Disk.Allocation {
 				groupScaling.Disk = &clouddatabasesv5.GroupScalingDisk{AllocationMb: core.Int64Ptr(int64(group.Disk.Allocation * nodeCount))}
 			}
 			if group.CPU != nil && group.CPU.Allocation != currentGroup.CPU.Allocation {
@@ -1950,7 +1950,7 @@ func resourceIBMDatabaseInstanceUpdate(context context.Context, d *schema.Resour
 				setDeploymentScalingGroupResponse, response, err := cloudDatabasesClient.SetDeploymentScalingGroup(setDeploymentScalingGroupOptions)
 
 				if response.StatusCode > 300 {
-					return diag.FromErr(err)
+					return diag.FromErr(fmt.Errorf("[ERROR] SetDeploymentScalingGroup (%s) failed %s\n%s", group.ID, err, response))
 				}
 
 				taskIDLink := *setDeploymentScalingGroupResponse.Task.ID

--- a/ibm/service/database/resource_ibm_database_elasticsearch_test.go
+++ b/ibm/service/database/resource_ibm_database_elasticsearch_test.go
@@ -76,6 +76,21 @@ func TestAccIBMDatabaseInstance_Elasticsearch_Basic(t *testing.T) {
 					resource.TestCheckResourceAttr(name, "connectionstrings.#", "1"),
 				),
 			},
+			{
+				Config: testAccCheckIBMDatabaseInstanceElasticsearchGroupMigration(databaseResourceGroup, testName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckIBMDatabaseInstanceExists(name, &databaseInstanceOne),
+					resource.TestCheckResourceAttr(name, "name", testName),
+					resource.TestCheckResourceAttr(name, "service", "databases-for-elasticsearch"),
+					resource.TestCheckResourceAttr(name, "plan", "standard"),
+					resource.TestCheckResourceAttr(name, "location", acc.IcdDbRegion),
+					resource.TestCheckResourceAttr(name, "groups.0.memory.0.allocation_mb", "3072"),
+					resource.TestCheckResourceAttr(name, "groups.0.disk.0.allocation_mb", "18432"),
+					resource.TestCheckResourceAttr(name, "whitelist.#", "0"),
+					resource.TestCheckResourceAttr(name, "users.#", "0"),
+					resource.TestCheckResourceAttr(name, "connectionstrings.#", "1"),
+				),
+			},
 			// {
 			// 	ResourceName:      name,
 			// 	ImportState:       true,
@@ -400,7 +415,7 @@ func testAccCheckIBMDatabaseInstanceElasticsearchFullyspecified(databaseResource
 			delete = "15m"
 		}
 	}
-	  
+
 				`, databaseResourceGroup, name, acc.IcdDbRegion)
 }
 
@@ -420,6 +435,41 @@ func testAccCheckIBMDatabaseInstanceElasticsearchReduced(databaseResourceGroup s
 		adminpassword                = "password12"
 		members_memory_allocation_mb = 3072
 		members_disk_allocation_mb   = 18432
+
+		timeouts {
+			create = "120m"
+			update = "120m"
+			delete = "15m"
+		}
+	}
+				`, databaseResourceGroup, name, acc.IcdDbRegion)
+}
+
+func testAccCheckIBMDatabaseInstanceElasticsearchGroupMigration(databaseResourceGroup string, name string) string {
+	return fmt.Sprintf(`
+	data "ibm_resource_group" "test_acc" {
+		is_default = true
+		# name = "%[1]s"
+	}
+
+	resource "ibm_database" "%[2]s" {
+		resource_group_id            = data.ibm_resource_group.test_acc.id
+		name                         = "%[2]s"
+		service                      = "databases-for-elasticsearch"
+		plan                         = "standard"
+		location                     = "%[3]s"
+		adminpassword                = "password12"
+
+		group {
+		  group_id = "member"
+
+		  memory {
+		    allocation_mb = 1024
+		  }
+		  disk {
+		    allocation_mb = 6144
+		  }
+		}
 
 		timeouts {
 			create = "120m"

--- a/ibm/service/database/resource_ibm_database_postgresql_test.go
+++ b/ibm/service/database/resource_ibm_database_postgresql_test.go
@@ -109,6 +109,22 @@ func TestAccIBMDatabaseInstancePostgresBasic(t *testing.T) {
 					resource.TestCheckResourceAttr(name, "tags.#", "1"),
 				),
 			},
+			{
+				Config: testAccCheckIBMDatabaseInstancePostgresGroupMigration(databaseResourceGroup, testName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckIBMDatabaseInstanceExists(name, &databaseInstanceOne),
+					resource.TestCheckResourceAttr(name, "name", testName),
+					resource.TestCheckResourceAttr(name, "service", "databases-for-postgresql"),
+					resource.TestCheckResourceAttr(name, "plan", "standard"),
+					resource.TestCheckResourceAttr(name, "location", acc.IcdDbRegion),
+					resource.TestCheckResourceAttr(name, "groups.0.memory.0.allocation_mb", "2048"),
+					resource.TestCheckResourceAttr(name, "groups.0.disk.0.allocation_mb", "14336"),
+					resource.TestCheckResourceAttr(name, "whitelist.#", "0"),
+					resource.TestCheckResourceAttr(name, "users.#", "0"),
+					resource.TestCheckResourceAttr(name, "connectionstrings.#", "1"),
+					resource.TestCheckResourceAttr(name, "tags.#", "1"),
+				),
+			},
 			// {
 			// 	ResourceName:      name,
 			// 	ImportState:       true,
@@ -568,6 +584,36 @@ func testAccCheckIBMDatabaseInstancePostgresReduced(databaseResourceGroup string
 		service_endpoints            = "public"
 		tags                         = ["one:two"]
 	  }
+				`, databaseResourceGroup, name, acc.IcdDbRegion)
+}
+
+func testAccCheckIBMDatabaseInstancePostgresGroupMigration(databaseResourceGroup string, name string) string {
+	return fmt.Sprintf(`
+	data "ibm_resource_group" "test_acc" {
+		name = "%[1]s"
+	}
+
+	resource "ibm_database" "%[2]s" {
+		resource_group_id            = data.ibm_resource_group.test_acc.id
+		name                         = "%[2]s"
+		service                      = "databases-for-postgresql"
+		plan                         = "standard"
+		location                     = "%[3]s"
+		adminpassword                = "password12"
+		service_endpoints            = "public"
+		tags                         = ["one:two"]
+
+		group {
+			group_id = "member"
+
+			memory {
+				allocation_mb = 1024
+			}
+			 disk {
+				allocation_mb = 7168
+			}
+		}
+	}
 				`, databaseResourceGroup, name, acc.IcdDbRegion)
 }
 

--- a/ibm/service/database/resource_ibm_database_redis_test.go
+++ b/ibm/service/database/resource_ibm_database_redis_test.go
@@ -69,6 +69,18 @@ func TestAccIBMDatabaseInstance_Redis_Basic(t *testing.T) {
 					resource.TestCheckResourceAttr(name, "whitelist.#", "0"),
 				),
 			},
+			{
+				Config: testAccCheckIBMDatabaseInstanceRedisGroupMigration(databaseResourceGroup, testName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "name", testName),
+					resource.TestCheckResourceAttr(name, "service", "databases-for-redis"),
+					resource.TestCheckResourceAttr(name, "plan", "standard"),
+					resource.TestCheckResourceAttr(name, "location", acc.IcdDbRegion),
+					resource.TestCheckResourceAttr(name, "groups.0.memory.0.allocation_mb", "2048"),
+					resource.TestCheckResourceAttr(name, "groups.0.disk.0.allocation_mb", "4096"),
+					resource.TestCheckResourceAttr(name, "whitelist.#", "0"),
+				),
+			},
 		},
 	})
 }
@@ -150,7 +162,7 @@ func testAccCheckIBMDatabaseInstanceRedisBasic(databaseResourceGroup string, nam
 		is_default = true
 		# name = "%[1]s"
 	  }
-	  
+
 	  resource "ibm_database" "%[2]s" {
 		resource_group_id            = data.ibm_resource_group.test_acc.id
 		name                         = "%[2]s"
@@ -173,9 +185,9 @@ func testAccCheckIBMDatabaseInstanceRedisFullyspecified(databaseResourceGroup st
 	data "ibm_resource_group" "test_acc" {
 		is_default = true
 		# name = "%[1]s"
-	  }
-	  
-	  resource "ibm_database" "%[2]s" {
+	}
+
+	resource "ibm_database" "%[2]s" {
 		resource_group_id            = data.ibm_resource_group.test_acc.id
 		name                         = "%[2]s"
 		service                      = "databases-for-redis"
@@ -192,7 +204,7 @@ func testAccCheckIBMDatabaseInstanceRedisFullyspecified(databaseResourceGroup st
 		  address     = "172.168.1.1/32"
 		  description = "desc"
 		}
-	  }
+	}
 				`, databaseResourceGroup, name, acc.IcdDbRegion)
 }
 
@@ -202,7 +214,7 @@ func testAccCheckIBMDatabaseInstanceRedisReduced(databaseResourceGroup string, n
 		is_default = true
 		# name = "%[1]s"
 	  }
-	  
+
 	  resource "ibm_database" "%[2]s" {
 		resource_group_id            = data.ibm_resource_group.test_acc.id
 		name                         = "%[2]s"
@@ -216,13 +228,41 @@ func testAccCheckIBMDatabaseInstanceRedisReduced(databaseResourceGroup string, n
 				`, databaseResourceGroup, name, acc.IcdDbRegion)
 }
 
+func testAccCheckIBMDatabaseInstanceRedisGroupMigration(databaseResourceGroup string, name string) string {
+	return fmt.Sprintf(`
+	data "ibm_resource_group" "test_acc" {
+		is_default = true
+	}
+
+	resource "ibm_database" "%[2]s" {
+		resource_group_id            = data.ibm_resource_group.test_acc.id
+		name                         = "%[2]s"
+		service                      = "databases-for-redis"
+		plan                         = "standard"
+		location                     = "%[3]s"
+		adminpassword                = "password12"
+
+		group {
+			group_id = "member"
+
+			memory {
+				allocation_mb = 1024
+			}
+			 disk {
+				allocation_mb = 2048
+			}
+	  }
+  }
+				`, databaseResourceGroup, name, acc.IcdDbRegion)
+}
+
 func testAccCheckIBMDatabaseInstanceRedisImport(databaseResourceGroup string, name string) string {
 	return fmt.Sprintf(`
 	data "ibm_resource_group" "test_acc" {
 		is_default = true
 		# name = "%[1]s"
 	  }
-	  
+
 	  resource "ibm_database" "%[2]s" {
 		resource_group_id = data.ibm_resource_group.test_acc.id
 		name              = "%[2]s"

